pkg/kubelet/kubelet.go

```diff
@@ -1671,6 +1672,32 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 	return true, "", ""
 }
 
+// HandleAdmissionFailure allows the kubelet to gracefully handle admission rejection, and, in some cases,
+// to allow admission of the pod despite its previous failure.
+func (kl *Kubelet) HandleAdmissionFailure(pod *v1.Pod, failureReasons []algorithm.PredicateFailureReason) (bool, []algorithm.PredicateFailureReason, error) {
+	if !kubetypes.IsCriticalPod(pod) {
+		return false, failureReasons, nil
+	}
+	// InsufficientResourceError is not a reason to reject a critical pod.
+	// Instead of rejecting, we free up resources to admit it, if no other reasons for rejection exist.
+	nonResourceReasons := []algorithm.PredicateFailureReason{}
+	resourceReasons := []predicates.InsufficientResourceError{}
```

Comment:
If there are "resourceReasons", those need to be passed to the preemption logic, so that it can preempt pods to alleviate those resources and admit the pod.
If there are "nonResourceReasons" then we cannot admit the pod.  While this is a yes/no decision, the predicate Admit function will return the first reason to the user as the "message" for why it chose not to admit the pod.  It would be misleading if a user was returned an InsufficientResourceError when attempting to admit a critical pod, since we are overriding those in a sense.  Instead, we want to return a "real" or "nonResourceReason" to them so they can correct it.